Add block visibility breakpoints support for dynamic configuration

ramonjd opened the pull request:
> [!IMPORTANT]
> This is an experimental PR for illustrative purposes only. If the approach is sound, it'll be broken up into smaller PRs for review and refactor.

Mostly vibe coded to demonstrate a flow for https://github.com/WordPress/gutenberg/issues/72502

### TODO

- [ ] The breakpoints should also trigger when changing the viewport in the dropdown, however `useViewPort` isn't looking at the iframe width, but the parent window. So somehow we need to know which device the user has selected, but it's [in the editor package](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/data/data-core-editor.md#L277), which can't be used in the block-editor package 🤔 Also very conveniently the values for mobile/tablet/etc are [unique in the drop down](https://github.com/WordPress/gutenberg/blob/try/hide-blocks-breakpoints/packages/block-editor/src/components/use-resize-canvas/index.js#L32). [Look at the use viewport match ones](https://github.com/WordPress/gutenberg/blob/try/hide-blocks-breakpoints/packages/compose/src/hooks/use-viewport-match/index.js#L22).
- [ ] Implement designs in https://github.com/WordPress/gutenberg/issues/72502#issuecomment-3611161242
- [ ] Split PR up for ease of reviewing 

https://github.com/user-attachments/assets/dccc06f8-a547-4127-8b80-13de1e03a438





<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Closes <!-- #ISSUE-NUMBER or URL -->

<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

|Before|After|
|-|-|
|<!-- Before screenshot here -->|<!-- After screenshot here -->|

ramonjd committed:
Add block visibility breakpoints support for dynamic configuration

ramonjd committed:
Combine menu

ramonjd committed:
Refactor block visibility breakpoints: remove menu item, streamline toolbar integration, and enhance modal functionality with improved breakpoint handling.

ramonjd committed:
Enhance block visibility breakpoints handling by adding a check for the existence of breakpoints before processing. Update paragraph block example to include a complete content attribute for better clarity.

ramonjd commented on docs/contributors/code/block-visibility-breakpoints-extensibility.md:
```diff
@@ -0,0 +1,186 @@
+# Block Visibility Breakpoints - Extensibility Considerations
```
To be reviewed. Any solution should leave open the requirement to have theme and/or user-defined breakpoints/widths

ramonjd replied:
Note: this document is speculative trash. It's just a planning/discussion document and will be deleted.

Don't rely on any single word written here.

ramonjd commented on lib/block-supports/block-visibility-breakpoints.php:
```diff
@@ -0,0 +1,104 @@
+<?php
```
This could be incorporated into the existing block-visibility.php file

ramonjd commented on packages/block-editor/src/components/block-visibility-breakpoints/modal-manager.js:
```diff
@@ -0,0 +1,56 @@
+		return () => listeners.delete( listener );
+	}, [] );
+
+	return state;
```
This was to get it to work and needs refactoring.

The modal slot should live outside the canvas so it's unaffected by the context menu mounting/unmounting.

ramonjd commented on packages/block-editor/src/components/block-visibility-breakpoints/modal.js:
```diff
@@ -0,0 +1,265 @@
+						checked={ hideEverywhere }
+						onChange={ handleHideEverywhereChange }
+					/>
+					<div className="block-editor-block-visibility-breakpoints-modal__breakpoints">
```
Lots of potential for visual/structural improvements here, e.g., using Stack components. Spacing is a bit weird.

ramonjd commented on packages/block-editor/src/components/provider/index.js:
```diff
@@ -111,6 +112,7 @@ export const ExperimentalBlockEditorProvider = withRegistryProvider(
 			<SlotFillProvider passthrough>
 				{ ! settings?.isPreviewMode && <KeyboardShortcuts.Register /> }
 				<BlockRefsProvider>{ props.children }</BlockRefsProvider>
+				<BlockVisibilityBreakpointsModalManager />
```
Is this the right place for the slot? I don't know.

ramonjd commented on lib/block-supports/block-visibility-breakpoints.php:
```diff
@@ -0,0 +1,104 @@
+				display: none !important;
+			}
+		}
+	';
```
Breakpoint values exist in a few places:

- https://github.com/WordPress/gutenberg/blob/651badeaea8b4b1fb046868485e9d8c3993a5fdf/packages/components/src/utils/breakpoint-values.js
- https://github.com/WordPress/gutenberg/blob/651badeaea8b4b1fb046868485e9d8c3993a5fdf/packages/base-styles/_breakpoints.scss
- https://github.com/WordPress/gutenberg/blob/651badeaea8b4b1fb046868485e9d8c3993a5fdf/packages/compose/src/hooks/use-viewport-match/index.js

ramonjd committed:
Refactor block visibility breakpoints modal to ensure that visibility states are only checked if all blocks share the same values. Update logic to handle empty block arrays and improve initial state calculations for breakpoints and "Hide everywhere" option.

ramonjd committed:
Enhance block visibility logic by adding checks for block visibility and breakpoint visibility. A block is now considered hidden if it is set to be hidden everywhere or if it has any visibility settings for mobile, tablet, or desktop breakpoints.

tellthemachines commented on docs/contributors/code/block-visibility-breakpoints-extensibility.md:
````diff
@@ -0,0 +1,186 @@
+}
+```
+
+This structure assumes exactly three semantic breakpoint ranges and doesn't support arbitrary breakpoint names.
````
Thought: if we were to consider supporting an arbitrary number of breakpoints to be defined with custom names, what might the editor UI to view different breakpoints evolve into?

<img width="265" height="305" alt="Image" src="https://github.com/user-attachments/assets/a9dafc22-1e2b-4c23-ae62-8c03d5261949" />

ramonjd replied:
True. Assuming we do allow `n` presets like we do with font sizes etc, they solve that using select boxes for items greater than `x`.

tellthemachines replied:
Or we could just move to resize handles everywhere instead of these fixed options 😄

ramonjd replied:
> Or we could just move to resize handles everywhere instead of these fixed options

Yeah, the hard-coded values seem pretty prescriptive. I reckon a lot of these types of issues will crawl out of their respective holes and demand solutions if Core ever lets folks add custom breakpoints.

jasmussen replied:
> Thought: if we were to consider supporting an arbitrary number of breakpoints to be defined with custom names, what might the editor UI to view different breakpoints evolve into?

This is not a direct answer, but contains some tangential thoughts that affect this interface:

* The Preview UI showing a desktop icon is not intuitive, if you want to preview in a new tab, it's not where you'd look.
* With collaborative and block comments underway, eventually we'll likely need a closer-to-industry-standard dropdown UI for things like "Suggesting".
* The button to "view the published post" that only appears if a post is published, honestly that button is useful all the time.
* There's a LOT of buttons in the top toolbar, even more when there are extension sidebars. Aside general cleanup and careful consideration, we should look at spacing density to optimize the space.
* The document title blob is useful to see what you're editing, but it's not intuitive as far as isolated editing navigation where I'd expect breadcrumbs, and honestly I'd expect to be able to change the title, or perhaps do post actions like delete, duplicat, by engaging with this control. Instead it just opens the command palette, which has a different invokation affordance inside the editor vs. outside.
* And finally, breakpoint editing, and major responsive tools are missing. The fact that we show desktop/mobile/tablet as purely preview tools, is a little misleading. In fact the resize handles you see on patterns being edited in isolation, is a more ergonomic tool for this.

All of that is me saying, the top toolbar deserves a vision and some near term steps towards it. I could see "Preview in new tab" and "View post" in a menu that also contained mode switches like Edit, Suggest, Preview. Where that surfaces clearer responsive tools needs some careful thought, but the benefit of each of those modes is each of them are context switches that can surface contextual tools.

ramonjd committed:
Add block visibility breakpoints constants and enhance modal functionality

Introduce constants for block visibility breakpoints to streamline the handling of responsive visibility settings. Update the block visibility breakpoints modal to utilize these constants, improving clarity and maintainability. Enhance error handling and ensure consistent state management across the modal and toolbar components.

ramonjd committed:
Remove block visibility breakpoints support and enhance block visibility logic. The `block-visibility-breakpoints.php` file has been deleted, and its functionality has been integrated into `block-visibility.php`, which now handles both "hide everywhere" and responsive breakpoint visibility. This streamlines the code and improves maintainability.

ramonjd committed:
Refactored block visibility checks to use a common function, reducing duplication and improving code organization.

ramonjd committed:
Update block visibility logic to use strict boolean checks for breakpoint visibility. Introduced a new test suite for validating block visibility behavior across various scenarios, ensuring accurate handling of visibility classes based on defined breakpoints.

ramonjd committed:
Refactor block visibility components to integrate modal functionality directly into menu items and toolbars. Removed the modal manager and slot system, simplifying state management for visibility breakpoints. Enhanced user interaction by allowing modals to open directly from toolbar buttons and menu items.

ramonjd committed:
Update block visibility breakpoints modal text and improve z-index handling. Changed the label from 'Hide everywhere' to 'Hide from published document' for clarity. Added z-index styling to ensure proper layering of the modal over other components.

ramonjd committed:
Remove block visibility breakpoints documentation file. This deletion eliminates outdated content regarding hardcoded breakpoints and extensibility considerations, streamlining the documentation for block visibility features.

ramonjd committed:
Refactor block-list styles to consolidate hidden block visibility rules. Combined multiple selectors for hidden blocks under flex layout into a single declaration for improved readability and maintainability.

ramonjd committed:
Added useRef and useEffect hooks to manage the visibility button's state, ensuring it remains in the toolbar if previously shown, even when no visibility is set. Updated component documentation for clarity on rendering conditions.

ramonjd committed:
Integrate success notices in block visibility breakpoints modal. Added notifications for hidden blocks and saved visibility settings, utilizing keyboard shortcuts for improved user feedback.

ramonjd committed:
e2e tests for block visibility breakpoints

tellthemachines reviewed (commented):
Quick note that breakpoint visibility isn't working in the editor; I tried hiding a block only on mobile and it disappeared altogether for all breakpoints.

<img width="720" height="613" alt="Image" src="https://github.com/user-attachments/assets/08b26e08-30dd-44eb-824a-747411f4dd76" />

It's working fine on the front end though.

The other thing I'm curious about is why we went with
```
visibility: hidden;
    overflow: hidden;
    height: 0;
    border: none !important;
    padding: 0 !important;
```
for hiding blocks in the editor instead of just `display: none`? I understand that choice was made when we added the option to hide blocks regardless of breakpoint.

ramonjd commented:
> I tried hiding a block only on mobile and it disappeared altogether for all breakpoints.

Yeah, lots to do still on this WIP branch. I'll take a look, thanks!

> instead of just display: none? I understand that choice was made when we added the option to hide blocks regardless of breakpoint.

Is the answer in the comment above the style?

```scss
	// Hidden blocks.
	// In order for the block toolbar to render correctly, blocks cannot be hidden.
	// Instead, use styles to visually hide blocks.
```

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/content.scss#L101-L104

ramonjd committed:
Using WP components

ramonjd committed:
Refactor block visibility logic in useBlockProps to improve clarity and functionality. Consolidated visibility settings retrieval and updated class application for hidden blocks based on overall visibility state.

ramonjd committed:
Extract logic for preview breakpoints to a separate hook.

ramonjd committed:
Refactor block visibility logic in isBlockHidden selector to simplify conditions. Now only checks if blockVisibility is false, as breakpoint visibility is managed via CSS classes.

ramonjd committed:
Enhance block visibility breakpoints modal with conditional rendering. Added explanatory text for hidden blocks based on visibility settings, improving user guidance on block management in the editor.

ramonjd committed:
List view icon will now show if the block is hidden everywhere or has any breakpoint visibility set.

ramonjd committed:
Add device type handling to block visibility logic

- Introduced `deviceTypeKey` to manage device type settings in block editor.
- Updated `usePreviewBreakpoint` hook to override viewport detection based on device type (Mobile, Tablet, Desktop).
- Enhanced `useBlockEditorSettings` to retrieve and utilize device type for improved block visibility management.

ramonjd commented on lib/block-supports/block-visibility.php:
```diff
@@ -19,14 +25,73 @@ function gutenberg_render_block_visibility_support( $block_content, $block ) {
 	remove_filter( 'render_block', 'wp_render_block_visibility_support' );
 }
 add_filter( 'render_block', 'gutenberg_render_block_visibility_support', 10, 2 );
+add_action( 'wp_enqueue_scripts', 'gutenberg_enqueue_block_visibility_breakpoints_styles' );
```
This would use the style engine enqueuing system eventually, just a shortcut for now

jasmussen commented:
Looks good at a glance. What happens if you deselect the block after hiding it? That's the most complex aspect to ensure a good behavior for.
